### PR TITLE
wallet: lock cs_main while accessing chainActive

### LINKED ISSUE #4684: Missing locking in wallet code
While working on #4680, we noticed some locking problems. The `cs_main` lock should be held while accessing `chainActive`, and it isn't in several places in the wallet code. There are also several places where the lock is correctly held but it's not obvious, so a call to `AssertLockHeld()` should be added.

### PR DESCRIPTION
**Pull request by LarryRuane**
Closes #4684

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -3535,6 +3537,7 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
 
     wtxNew.fTimeReceivedIsTxTime = true;
     wtxNew.BindWallet(this);
+    LOCK(cs_main);
```
Comment (daira) [inline src/wallet/wallet.cpp:3540]: This implies that the `LOCK2(cs_main, cs_wallet);` at line 3581 is redundant and should be changed to
```
// cs_main already taken above
LOCK(cs_wallet);
```

**Review by daira — APPROVED** [review]
utACK with a minor comment.

**Comment by LarryRuane** [comment]
Force pushed @daira's review suggestion, good catch! ([diff](https://github.com/zcash/zcash/compare/d85275784d8cedfe28a2ee1af5d81bb057f7e805..fcfac7b7fc622cc6349bcd017aa00d03d0c2cc1b))

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (LarryRuane) [reply src/wallet/wallet.cpp:3540]: Darn, I forgot to include the comment in the force-push I just did; do you think that's worth adding?

**Comment by LarryRuane** [comment]
Another force-push, forgot to add the comment @daira suggested ([diff](https://github.com/zcash/zcash/compare/fcfac7b7fc622cc6349bcd017aa00d03d0c2cc1b..3261ba41bfcf5985a44fc1ed9be4ec69887cb626))

**Comment by LarryRuane** [comment]
@zkbot try

**Review by daira — APPROVED** [review]
utACK

**Comment by daira** [comment]
Intermittent ccache-related problem.

**Comment by ebfull** [comment]
@zkbot r+

**Pull request merged**